vmm: use gsi bitmap to record the GSI used

up2wing opened the pull request:
vmm: use gsi bitmap to record the GSI used
    
KVM use KVM_MAX_IRQ_ROUTES(4096) as the max gsi which vm can use.
If we use gsi without restraint, it will eventually exceed that  value and fail.

Add a bitmap to track the gsi enable/disable, so as to maxmize the use the gsi in kvm.

russell-islam commented on vm-allocator/src/gsi.rs:
```diff
@@ -29,12 +33,64 @@ impl GsiApic {
 }
 
+#[cfg(target_arch = "x86_64")]
+pub struct GsiBitmap {
```
Why don't you use the bitmap provided by vm-memory crate?

up2wing replied:
The Bitmap in vm-memory seems a bit heavy in our this scenario. All we want is a bitmap which only has enable/disable one bit function.
Also, our bitmap code is very tiny which has only 50 lines :)

russell-islam replied:
It's not too heavy as we have used it before in MSHV hypervisor implementations. There are multiple implementations available. Please take a look if that does not work for you, feel free to add. But adding more code is also a maintenance burden,

russell-islam replied:
Yes, I agree. But instead of adding more Bitmap related code, I would prefer using/enhancing exiting ones.

up2wing replied:
> https://crates.io/crates/bit-vec seems to be the popular choice.

Thx, I will look into that.

up2wing replied:
Ahhhh, sorry, my fault. I set the bitmap incorrectly when change GsiBitmap to bit-vec, which I thought is a simple change and didn't check carefully.
Now it seems work fine.

rbradford commented on vm-device/src/interrupt/mod.rs:
```diff
@@ -68,6 +68,11 @@ pub type Result<T> = std::io::Result<T>;
+pub const IRQ_KEEP_UNMASKED: usize = 1 << 1;
+pub const IRQ_MASKED_TO_UNMASKED: usize = 1 << 2;
+pub const IRQ_UNMASKED_TO_MASKED: usize = 1 << 3;
+
```
This could be an `enum` but it seems a bit of weird way to implement this - why not store the previous mask value and check if the mask request has changed and then update?

up2wing replied:
Yes, using enum here makes the code look neater, but we need to implement BitOr and PartialEq, which increases the complexity of the code and I think may be more error-prone.
Storing the previous mask value is a good idea, however the mask state is a state of gsi, and we need to store a HashMap or something like that in InterruptSourceGroup, so I suppose the code in the PR is the more simple way.
Could we keep this usize code plz? :)

up2wing commented on vm-allocator/src/system.rs:
```diff
@@ -91,6 +91,16 @@ impl SystemAllocator {
+    }
+
+    /// Enable GSI bitmap
+    pub fn enable_gsi_bit(&mut self, gsi: u32) {
```
thx, done.

up2wing commented on vmm/src/interrupt.rs:
```diff
@@ -5,46 +5,74 @@
 
+    pub fn get_gsi(&self) -> Option<u32> {
+        let mut gsi = self.gsi.load(Ordering::SeqCst);
+        if u32::MAX == gsi {
```
done, thx.

rbradford commented on vm-allocator/src/gsi.rs:
```diff
@@ -67,17 +73,36 @@ impl GsiAllocator {
+    }
+
+    /// Enable gsi bit
+    pub fn enable_gsi_bit(&mut self, gsi: u32) {
```
It's a very strange layering issue to have this API here - it would be better to have enable/disable.

up2wing replied:
fixed, thx a lot.

up2wing commented:
> @up2wing Please rebase. 
Thx, rebased. 

> Also, have you addressed all comments from @rbradford?
I suppose yes.

I'm very sorry that I haven't spent much time on the community due to being too busy with work recently. Hope to have more time to this great work in the future.

likebreath commented on vm-allocator/src/gsi.rs:
```diff
@@ -67,17 +73,36 @@ impl GsiAllocator {
+    pub fn enable_gsi_bit(&mut self, gsi: u32) {
+        self.gsi_bitmap.lock().unwrap().set(gsi as usize, true);
+    }
+
```
Can you please explain why this function is needed?
